core/trino-main/src/main/java/io/trino/operator/exchange/PartitioningExchanger.java

```diff
@@ -77,7 +77,9 @@ private synchronized void partitionPage(Page page, Page partitionPage)
             }
             Page pageSplit;
             if (partitionSize == page.getPositionCount()) {
-                pageSplit = page; // entire page will be sent to this partition, no copies necessary
+                // entire page will be sent to this partition, just compact the page to avoid over-retaining memory
```

Comment:
I don't think we should compact data eagerly here. This pages will usually land in `PagesIndex` anyway which will compact pages if eagerly, see `io.trino.operator.PagesIndex#compact`

Reply:
Having `PagesIndex#compact` handle this case will actually prevent `HashBuilderOperator` from spilling to disk based on `HashBuilderOperator#startMemoryRevoke()` deciding that reductions in retained size of 20% or better upon compaction are sufficient to avoid starting to spill.